microsoft/testsuites/dpdk/common.py

```diff
@@ -17,6 +22,266 @@
+    def get_installed_version(self) -> VersionInfo:
+        raise NotImplementedError(f"get_installed_version {self._err_msg}")
+
+    # run the defined setup and installation steps.
+    def do_installation(self, required_version: Optional[VersionInfo] = None) -> None:
+        self._setup_node()
+        if self._check_if_installed() and required_version is not None:
+            if self.get_installed_version() == required_version:
+                return
+            else:
+                self._clean_previous_installation()
+        self._setup_installation()
+        self._configure_installation()
+        self._run_installation()
+
+    def __init__(
+        self, node: Node, os_dependencies: Optional[InstallDependencies] = None
+    ) -> None:
+        self._node = node
+        if not isinstance(self._node.os, Posix):
+            raise UnsupportedDistroException(
+                self._node.os, "Installer parent class requires Posix OS."
+            )
+        self._os: Posix = self._node.os
+        self._package_manager_extra_args: List[str] = []
+        self._os_dependencies = os_dependencies
+
+
+# Base class for package manager installation
+class PackageManagerInstall(Installer):
+    def __init__(self, node: Node, os_dependencies: InstallDependencies) -> None:
+        super().__init__(node, os_dependencies)
+
+    # define no-ops for source setup functions
+    def _setup_installation(self) -> None:
+        ...
+
+    def _configure_installation(self) -> None:
+        ...
+
+    # uninstall from the package manager
+    def _clean_previous_installation(self) -> None:
+        if not isinstance(self._os, Posix):
+            return
+        if self._os_dependencies is not None:
+            for os_package_check in self._os_dependencies.requirements:
+                if os_package_check.check_node_os(self._os):
+                    self._os.uninstall_packages(os_package_check.packages)
+
+    # verify packages on the node have been installed by
+    # the package manager
+    def _check_if_installed(self) -> bool:
+        if not isinstance(self._os, Posix):
+            return False
+        # WARNING: Don't use this for long lists of packages.
+        # For dpdk, pkg-manager install is only for 'dpdk' and 'dpdk-dev'
+        # This will take too long if it's more than a few packages.
+        if self._os_dependencies is not None:
+            for os_package_check in self._os_dependencies.requirements:
+                if os_package_check.check_node_os(self._os):
+                    for pkg in os_package_check.packages:
+                        if not self._os.package_exists(pkg):
+                            return False
+        return True
+
+    def _run_installation(self) -> None:
+        # define to handle any special special cases
+        self._setup_node()
+        # and install everything
+        self._install_dependencies()
+
+
+# base class for source installs
+class SourceInstall(Installer):
```

Comment:
The sourceinstall class has no special implementation for source installation, it looks not needed.

Reply:
I'm going to leave it, even though it doesn't do much.
- It sets _checkout_path
- It serves as a common class for checking isinstance(installer, SourceInstall)

I'm going to re-use this code for rdma-core as well, so having common parent classes is going to be useful.

Reply:
1. It's working path, not checkout path. The "checkout" is for GitInstall, but it's a "source" path for TarInstall. So it still presents a "working" path. Also the logic is simple, just save the node working path. It doesn't need to have a base class for one line assignment.
2. You can think about design ahead, but not do too much. When you implement rdma, you can refactor at that moment.

Reply:
Ok, I'll spend the time to remove this code and then add it again later when it's needed.

Reply:
ok fixed